packages/block-editor/src/components/link-control/use-remote-url-data.js

```diff
@@ -39,19 +39,27 @@ function useRemoteUrlData( url ) {
 		isFetching: false,
 	} );
 
-	const { fetchRemoteUrlData } = useSelect( ( select ) => {
+	const { fetchRemoteUrlData, baseUrl } = useSelect( ( select ) => {
 		const { getSettings } = select( blockEditorStore );
 		return {
+			baseUrl: getSettings().__experimentalBaseUrl,
 			fetchRemoteUrlData: getSettings().__experimentalFetchRemoteUrlData,
```

Comment:
So If I'm not wrong, this feature requires two block-editor settings, It feels a bit too much for just previews tbh. Can't we introduce the bail out behavior inside `__experimentalFetchRemoteUrlData` somehow?

Reply:
That was my first thought. Should that _need_ to know about the concept of an internal URL though? 

I can look into that later on today.

Reply:
Maybe the block-editor doesn't need to know about "fetchRemoteURL" entirely? I'm just thinking out loud but maybe the setting should be `renderURLPreview` (render function)

Reply:
That is an interesting option.

Reply:
But does it mean that the users have to implement the loading and error status/style by themselves?

How about instead of accepting an async function (fetchRemoteUrlData), or a render function (renderUrlPreview), we simply accept a React hook (useRemoteUrlData) as the setting value? The hook takes a remote `url` as the argument, and return `richData` and `isFetching`, just like the one we've already created.

```js
useRemoteUrlData( url ) {
  // ...

  return { richData, isFetching };
}
```

This way, the users don't have to implement the loading style, and they can also skip fetching internal URLs if the `url` matches a certain pattern. There are no intermediate loading status like we have in async function since that everything can return synchronously. In the future, we can also handle errors or support other progressive enhancements directly in the hook.

Reply:
If we can do that then it sounds great @kevin940726. 

If we do this then we should make this setting name "remote URL" agnostic (eg: `handleRichPreview` or something) so that in future we can use it to handle both:

1. External URLs
2. **_Internal_** URLs.

For now, the hook can just handle remote URLs and bail out for internal ones.

I assume the component can just call the hook. It doesn't care that it comes in via a block editor setting so we should be good to go with this idea.

How does that sound @youknowriad?

Reply:
so it would be something like:

```
const { useRemoteURLData } = useSelect( select => select( 'block-editor' ).getSettings(), [] );
const { richData, isLoading } = useRemoteURLData();
```

this kind of breaks the rules of hooks, because that hook is not necessarily defined and can be updated over time to call different built-in hooks.

I do think it could be nice, but maybe a bit fragile and a bit weird as well (a react hook as a setting). 

WDYT

Reply:
@youknowriad What you have is (I think) how @kevin940726 and I imagined it.

> because that hook is not necessarily defined

That's probably the biggest problem for me. It's likely to only be defined in the Editor and not the Block Editor.

Reply:
How about we change the setting to `fetchRichData`.

This accepts a `url` much as `__experimentalFetchRemoteUrlData` does now. But within that function we decide which path to follow based on whether the `url` is internal or not.

Initially the implementation would be something like:

```js
function __experimentalFetchRichData(url) {
    if (url.includes(baseUrl)) {
        return Promise.resolve({})'
    }

    // Now do the same as `__experimentalFetchRemoteUrlData` does currently
    // to handle external URL
}
```

Then in the future we could bolt on the ability to fetch rich data about an _internal_ URL very easily. 

This allows us to keep the current hook implementation the same whilst also making sure that `block-editor` has no idea about WP specific stuff.

Reply:
See https://github.com/WordPress/gutenberg/pull/32658

Reply:
What I was thinking is to split `<LinkPreview>` into 2 different components, one is the original `LinkPreview`, and the other one is the `LinkRichPreview` component. We can skip rendering `<LinkRichPreview>` if there's no `useRemoteURLData` (or any other name) so that we don't break the rules of hooks. And I would imagine the setting can't be updated once it's rendered so that it won't break the other rule of hooks either.

But I agree that it's a bit weird though. Maybe render function (or just a component) is a better way to go, but would need to update `LinkPreview` to make it work.